From c67e933016ee2142fdeed99de54b2b9e436d2fc6 Mon Sep 17 00:00:00 2001
From: Siye Liu <siliu@microsoft.com>
Date: Wed, 30 Oct 2024 05:33:19 +0000
Subject: [PATCH] Fix crash in |InputMethodController::ExtendSelectionAndDelete|

The crash is introduced by this CL[1]. In |ExtendSelectionAndDelete|
layout can be dirty again after |SetSelectionOffsets| call. We should
make sure that layout is clean before getting visible selection.

[1]: https://chromium-review.googlesource.com/c/chromium/src/+/5472192

Bug: 336992377
Change-Id: Ib165278ca332f3511de635bd328c600675fa18ab
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5963237
Commit-Queue: Koji Ishii <kojii@chromium.org>
Auto-Submit: Siye Liu <siliu@microsoft.com>
Reviewed-by: Koji Ishii <kojii@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1375679}
---

--- a/third_party/blink/renderer/core/editing/ime/input_method_controller.cc
+++ b/third_party/blink/renderer/core/editing/ime/input_method_controller.cc
@@ -1475,6 +1475,9 @@
     if (before == 0)
       break;
     ++before;
+    // TODO(editing-dev): The use of UpdateStyleAndLayout
+    // needs to be audited.  see http://crbug.com/590369 for more details.
+    GetDocument().UpdateStyleAndLayout(DocumentUpdateReason::kEditing);
   } while (
       GetFrame().Selection().ComputeVisibleSelectionInDOMTree().Start() ==
           GetFrame().Selection().ComputeVisibleSelectionInDOMTree().End() &&
